src/shogun/classifier/svm/NewtonSVM.cpp

```diff
@@ -69,63 +71,45 @@ bool CNewtonSVM::train_machine(CFeatures* data)
 
 	ASSERT(num_vec==train_labels.vlen)
 
-	float64_t* weights = SG_CALLOC(float64_t, x_d+1);
-	float64_t* out=SG_MALLOC(float64_t, x_n);
-	SGVector<float64_t>::fill_vector(out, x_n, 1.0);
-
-	int32_t *sv=SG_MALLOC(int32_t, x_n), size_sv=0, iter=0;
-	float64_t obj, *grad=SG_MALLOC(float64_t, x_d+1);
-	float64_t t;
-
-	auto pb = SG_PROGRESS(range(num_iter));
-	while (iter <= num_iter)
-	{
-		COMPUTATION_CONTROLLERS
-		iter++;
-
-		obj_fun_linear(weights, out, &obj, sv, &size_sv, grad);
-
-#ifdef DEBUG_NEWTON
-		SG_PRINT("fun linear passed !\n")
-		SG_PRINT("Obj =%f\n", obj)
-		SG_PRINT("Grad=\n")
-
-		for (int32_t i=0; i<x_d+1; i++)
-			SG_PRINT("grad[%d]=%.16g\n", i, grad[i])
-		SG_PRINT("SV=\n")
+	SGVector<float64_t>weights(x_d);
+	set_w(weights);
+	out = SGVector<float64_t>(x_n);
+	out.set_const(1.0);
+	weights.set_const(0.0);
+	
+	sv = SGVector<int32_t>(x_n);
+	sv.set_const(0);
+	grad = SGVector<float64_t>(x_d+1);
+	grad.set_const(0.0);
+}
 
-		for (int32_t i=0; i<size_sv; i++)
-			SG_PRINT("sv[%d]=%d\n", i, sv[i])
-#endif
+void CNewtonSVM::iteration()
+{
+	obj_fun_linear();
+	SGVector<float64_t> weights = get_w();
 
 		SGVector<float64_t> sgv;
-		float64_t* Xsv = SG_MALLOC(float64_t, x_d*size_sv);
+		SGMatrix<float64_t> Xsv(x_d, size_sv);
```

Comment:
Question: Can we re-use this memory across iterations? So we dont have to allocate-free-allocate-free it all the time? What is worse is actually that we set it to 0 in a loop everytime it is allocated. Could maybe be a member element and then be re-used?

Reply:
actually size_sv changes every iteration so the size of memory is not constant over here

Reply:
I see ... ok!

Reply:
ah but is it getting smaller or larger?
If just ever getting smaller, we could use resize .... which is might be more efficient?

Reply:
it gets smaller than bigger... its better like this i think

Reply:
OK, agreed